File: web/src/api/incidents.ts
import type {
  ActionStatus,
  DecisionAction,
  EvidenceDirection,
  IncidentActionState,
  IncidentConfidence,
  IncidentConfidenceStatus,
  IncidentEvidenceValue,
  IncidentFeedItem,
  IncidentFeedResponse,
  IncidentSeverity,
  IncidentState,
  VerdictClass,
} from '@/contracts/types';

const STATES = new Set<IncidentState>(['OPEN', 'MITIGATING', 'MONITORING', 'RESOLVED']);
const SEVERITIES = new Set<IncidentSeverity>(['CRITICAL', 'HIGH', 'MEDIUM', 'LOW']);
const VERDICTS = new Set<VerdictClass>([
  'EXPECTED_EVENT',
  'ATTACK',
  'OPERATIONAL_FAULT',
  'CODE_CONFIG_FAULT',
  'COMBINATION',
]);
const DECISIONS = new Set<DecisionAction>([
  'SUPPRESS',
  'ALERT',
  'ACT',
  'ESCALATE_TO_HUMAN',
  'AUTO_CONTAIN_THEN_ESCALATE',
]);
const EFFECTS = new Set<ActionStatus>([
  'SIMULATED',
  'APPLIED',
  'VERIFIED',
  'REVERTED',
  'FAILED',
]);
const DIRECTIONS = new Set<EvidenceDirection>([
  'ABOVE_BASELINE',
  'BELOW_BASELINE',
  'AT_BASELINE',
]);

function record(value: unknown, name: string): Record<string, unknown> {
  if (value === null || typeof value !== 'object' || Array.isArray(value)) {
    throw new Error(`${name} must be an object`);
  }
  return value as Record<string, unknown>;
}

function text(value: unknown, name: string): string {
  if (typeof value !== 'string' || value.trim().length === 0) {
    throw new Error(`${name} must be non-empty text`);
  }
  return value;
}

function time(value: unknown, name: string): string {
  const rendered = text(value, name);
  if (!/(?:Z|\+00:00)$/.test(rendered) || !Number.isFinite(Date.parse(rendered))) {
    throw new Error(`${name} must be a UTC timestamp`);
  }
  return rendered;
}

function finite(value: unknown, name: string): number {
  if (typeof value !== 'number' || !Number.isFinite(value)) {
    throw new Error(`${name} must be finite`);
  }
  return value;
}

function optionalText(value: unknown, name: string): string | null {
  return value === null ? null : text(value, name);
}

function member<T extends string>(value: unknown, values: Set<T>, name: string): T {
  if (typeof value !== 'string' || !values.has(value as T)) {
    throw new Error(`${name} is unknown`);
  }
  return value as T;
}

function parseEvidence(value: unknown): IncidentEvidenceValue {
  const body = record(value, 'incident evidence');
  const direction = member(body.direction, DIRECTIONS, 'incident evidence direction');
  const measured = finite(body.value, 'incident evidence value');
  const baseline = finite(body.baseline, 'incident evidence baseline');
  if (
    (measured > baseline && direction !== 'ABOVE_BASELINE') ||
    (measured < baseline && direction !== 'BELOW_BASELINE') ||
    (measured === baseline && direction !== 'AT_BASELINE')
  ) {
    throw new Error('incident evidence direction contradicts its values');
  }
  return {
    feature: text(body.feature, 'incident evidence feature'),
    value: measured,
    baseline,
    direction,
    note: text(body.note, 'incident evidence note'),
  };
}

function parseAction(value: unknown): IncidentActionState {
  const body = record(value, 'incident action');
  return {
    decision_action: member(body.decision_action, DECISIONS, 'incident decision action'),
    effect_status:
      body.effect_status === null
        ? null
        : member(body.effect_status, EFFECTS, 'incident effect status'),
    detail: text(body.detail, 'incident action detail'),
  };
}

function parseConfidence(value: unknown): IncidentConfidence {
  const body = record(value, 'incident confidence');
  const status = member(
    body.status,
    new Set<IncidentConfidenceStatus>(['calibrated', 'insufficient']),
    'incident confidence status',
  );
  const confidence = body.value === null ? null : finite(body.value, 'incident confidence');
  if (confidence !== null && (confidence < 0 || confidence > 1)) {
    throw new Error('incident confidence must be within [0, 1]');
  }
  if (
    (status === 'calibrated' && confidence === null) ||
    (status === 'insufficient' && confidence !== null)
  ) {
    throw new Error('incident confidence value contradicts its status');
  }
  return {
    status,
    value: confidence,
    note: text(body.note, 'incident confidence note'),
  };
}

function parseItem(value: unknown): IncidentFeedItem {
  const body = record(value, 'incident feed item');
  if (!Array.isArray(body.services) || body.services.length === 0) {
    throw new Error('incident services must be non-empty');
  }
  if (!Array.isArray(body.evidence) || body.evidence.length > 2) {
    throw new Error('incident evidence must contain at most two values');
  }
  const evidence: IncidentFeedItem['evidence'] =
    body.evidence.length === 0
      ? []
      : body.evidence.length === 1
        ? [parseEvidence(body.evidence[0])]
        : [parseEvidence(body.evidence[0]), parseEvidence(body.evidence[1])];
  const services: [string, ...string[]] = [
    text(body.services[0], 'incident service'),
    ...body.services.slice(1).map((service) => text(service, 'incident service')),
  ];
  if (new Set(services).size !== services.length) {
    throw new Error('incident services must be unique');
  }
  const origin = optionalText(body.origin_service, 'incident origin');
  if (origin !== null && !services.includes(origin)) {
    throw new Error('incident origin must be one of its services');
  }
  const verdict =
    body.verdict_class === null
      ? null
      : member(body.verdict_class, VERDICTS, 'incident verdict');
  const action = parseAction(body.action);
  const muted = body.muted;
  if (typeof muted !== 'boolean') throw new Error('incident muted state must be boolean');
  const explanation = optionalText(body.explanation, 'incident explanation');
  if (muted !== (explanation !== null)) {
    throw new Error('muted incident and explanation must agree');
  }
  if (muted && (verdict !== 'EXPECTED_EVENT' || action.decision_action !== 'SUPPRESS')) {
    throw new Error('only a suppressed expected event may be muted');
  }
  if (body.honesty !== 'REAL' && body.honesty !== 'SIMULATED') {
    throw new Error('incident honesty is unknown');
  }
  return {
    incident_id: text(body.incident_id, 'incident id'),
    opened_at: time(body.opened_at, 'incident opening'),
    updated_at: time(body.updated_at, 'incident update'),
    state: member(body.state, STATES, 'incident state'),
    severity: member(body.severity, SEVERITIES, 'incident severity'),
    services,
    origin_service: origin,
    verdict_class: verdict,
    reason: text(body.reason, 'incident reason'),
    evidence,
    action,
    confidence: parseConfidence(body.confidence),
    muted,
    explanation,
    honesty: body.honesty,
  };
}

export function parseIncidentFeedResponse(value: unknown): IncidentFeedResponse {
  const body = record(value, 'incident feed response');
  if (body.status !== 'ready' && body.status !== 'degraded') {
    throw new Error('incident feed status is unknown');
  }
  if (!Array.isArray(body.incidents)) throw new Error('incidents must be an array');
  if (
    typeof body.limit !== 'number' ||
    !Number.isInteger(body.limit) ||
    body.limit < 1 ||
    body.limit > 50
  ) {
    throw new Error('incident feed limit must be within [1, 50]');
  }
  const incidents = body.incidents.map(parseItem);
  if (
    typeof body.count !== 'number' ||
    !Number.isInteger(body.count) ||
    body.count !== incidents.length ||
    body.count > body.limit
  ) {
    throw new Error('incident feed count contradicts its items or limit');
  }
  const detail = body.detail === null ? null : text(body.detail, 'incident feed detail');
  if (
    (body.status === 'ready' && detail !== null) ||
    (body.status === 'degraded' && (detail === null || incidents.length > 0))
  ) {
    throw new Error('incident feed status contradicts its detail or items');
  }
  return {
    status: body.status,
    incidents,
    count: body.count,
    limit: body.limit,
    detail,
  };
}

export async function fetchIncidents(signal?: AbortSignal): Promise<IncidentFeedResponse> {
  const response = await fetch('/api/incidents?limit=20', signal ? { signal } : {});
  if (response.status !== 200 && response.status !== 503) {
    throw new Error(`incident request failed: ${response.status}`);
  }
  return parseIncidentFeedResponse(await response.json());
}

File: platform/api/incidents.py
"""Durable live-incident snapshots and post-write invalidation."""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Literal, Protocol

from pydantic import ValidationError

from api.causal_graph import build_causal_graph, causal_graph_record
from api.incident_detail import build_incident_detail, incident_detail_record
from api.security import build_security_snapshot, security_record
from api.stream import snapshot_invalidation
from common.config import TopologyConfig
from common.storage import (
    IncidentDetailRecord,
    IncidentGraphRecord,
    IncidentRecord,
    IncidentSecurityRecord,
)
from contracts import (
    ActionControlSnapshot,
    AuditEntry,
    DecisionAction,
    DecompFrame,
    IncidentActionState,
    IncidentConfidence,
    IncidentConfidenceStatus,
    IncidentEvidenceValue,
    IncidentFeedItem,
    IncidentFeedResponse,
    SecurityCohort,
    SecurityMeasurement,
    SnapshotInvalidation,
    SnapshotResource,
    SymptomEpisode,
    VerdictClass,
)
from decision import IncidentOutcome

MAX_INCIDENTS = 50
DEFAULT_INCIDENTS = 20


class IncidentFeedReader(Protocol):
    """The bounded runtime-store read used by the gateway."""

    async def list_incidents(self, *, limit: int) -> tuple[IncidentRecord, ...]: ...


class IncidentFeedStore(Protocol):
    """The durable write used by a live decision-output publisher."""

    async def put_incident_bundle(
        self,
        record: IncidentRecord,
        graph: IncidentGraphRecord,
        detail: IncidentDetailRecord,
        security: IncidentSecurityRecord,
        action_control: ActionControlSnapshot | None = None,
    ) -> bool: ...


class InvalidationBroker(Protocol):
    """The in-process fan-out seam owned by the gateway."""

    def publish(self, event: SnapshotInvalidation) -> int: ...


class IncidentFeedDataError(ValueError):
    """A stored runtime row could not support the public contract."""


@dataclass(frozen=True, slots=True)
class IncidentPublishResult:
    """What one decision-output persistence attempt changed."""

    item: IncidentFeedItem
    persisted: bool


class IncidentFeedPublisher:
    """Persist a decision outcome, then and only then invalidate its snapshot."""

    def __init__(self, *, store: IncidentFeedStore, broker: InvalidationBroker) -> None:
        self._store = store
        self._broker = broker

    async def persist(
        self,
        outcome: IncidentOutcome,
        *,
        episodes: tuple[SymptomEpisode, ...],
        topology: TopologyConfig,
        dependency_signal_prefix: str,
        honesty: Literal["REAL", "SIMULATED"],
        stimulus_honesty: Literal["REAL", "SIMULATED"],
        mode: Literal["LIVE", "REPLAY"] = "LIVE",
        capture_id: str | None = None,
        seed: int | None = None,
        decomp_frames: tuple[DecompFrame, ...] = (),
        audit_entries: tuple[AuditEntry, ...] = (),
        decomposition_truncated: bool = False,
        calibrated_confidence: float | None = None,
        calibration_note: str | None = None,
        explained_event: str | None = None,
        action_control: ActionControlSnapshot | None = None,
        security_cohorts: tuple[SecurityCohort, ...] = (),
        protected_cohort_integrity: SecurityMeasurement | None = None,
    ) -> IncidentPublishResult:
        """Write a newer incident revision and signal browsers after commit."""
        item = build_incident_feed_item(
            outcome,
            honesty=honesty,
            calibrated_confidence=calibrated_confidence,
            calibration_note=calibration_note,
            explained_event=explained_event,
        )
        graph = build_causal_graph(
            outcome,
            episodes=episodes,
            topology=topology,
            dependency_signal_prefix=dependency_signal_prefix,
            honesty=honesty,
        )
        detail = build_incident_detail(
            outcome,
            episodes=episodes,
            graph=graph,
            decomp_frames=decomp_frames,
            audit_entries=audit_entries,
            telemetry_honesty=honesty,
            stimulus_honesty=stimulus_honesty,
            mode=mode,
            capture_id=capture_id,
            seed=seed,
            calibrated_confidence=calibrated_confidence,
            calibration_note=calibration_note,
            decomposition_truncated=decomposition_truncated,
        )
        security = build_security_snapshot(
            outcome,
            episodes=episodes,
            decomposition=detail.decomposition,
            honesty=honesty,
            cohorts=security_cohorts,
            action_control=action_control,
            protected_cohort_integrity=protected_cohort_integrity,
        )
        persisted = await self._store.put_incident_bundle(
            incident_record(item),
            causal_graph_record(graph),
            incident_detail_record(detail),
            security_record(security),
            action_control,
        )
        if persisted:
            resources = (
                (
                    SnapshotResource.INCIDENTS,
                    SnapshotResource.ACTIONS,
                    SnapshotResource.SECURITY,
                )
                if action_control is not None
                else (SnapshotResource.INCIDENTS, SnapshotResource.SECURITY)
            )
            self._broker.publish(snapshot_invalidation(*resources))
        return IncidentPublishResult(item=item, persisted=persisted)


def build_incident_feed_item(
    outcome: IncidentOutcome,
    *,
    honesty: Literal["REAL", "SIMULATED"],
    calibrated_confidence: float | None = None,
    calibration_note: str | None = None,
    explained_event: str | None = None,
) -> IncidentFeedItem:
    """Project one full decision outcome into the compact public card."""
    verdict = outcome.verdict
    if explained_event is not None and (
        verdict is None
        or verdict.verdict_class is not VerdictClass.EXPECTED_EVENT
        or outcome.decision.action is not DecisionAction.SUPPRESS
    ):
        raise ValueError("an explained event requires a suppressed EXPECTED_EVENT outcome")
    confidence = (
        IncidentConfidence(
            status=IncidentConfidenceStatus.INSUFFICIENT,
            value=None,
            note=(
                _line(calibration_note)
                if calibration_note is not None
                else "Runtime rule confidence is not calibrated, so no confidence is displayed."
            ),
        )
        if calibrated_confidence is None
        else IncidentConfidence(
            status=IncidentConfidenceStatus.CALIBRATED,
            value=calibrated_confidence,
            note=_line(
                calibration_note
                if calibration_note is not None
                else "A calibration note is required for calibrated confidence."
            ),
        )
    )
    if calibrated_confidence is not None and calibration_note is None:
        raise ValueError("calibrated confidence requires a calibration note")

    evidence: tuple[IncidentEvidenceValue, ...] = ()
    if verdict is not None:
        strongest = sorted(
            verdict.evidence,
            key=lambda item: (-item.contribution, item.feature),
        )[:2]
        evidence = tuple(
            IncidentEvidenceValue(
                feature=item.feature,
                value=item.value,
                baseline=item.baseline,
                direction=item.direction,
                note=_line(item.note),
            )
            for item in strongest
        )
    incident = outcome.incident
    services = tuple(dict.fromkeys((*incident.services, *incident.implicated_services)))
    return IncidentFeedItem(
        incident_id=incident.incident_id,
        opened_at=incident.opened_ts,
        updated_at=outcome.decision.ts,
        state=incident.state,
        severity=incident.severity,
        services=services,
        origin_service=incident.origin_service,
        verdict_class=verdict.verdict_class if verdict is not None else None,
        reason=_line(outcome.decision.reason),
        evidence=evidence,
        action=IncidentActionState(
            decision_action=outcome.decision.action,
            effect_status=None,
            detail=_action_detail(outcome.decision.action),
        ),
        confidence=confidence,
        muted=explained_event is not None,
        explanation=(
            f"Explained by {_line(explained_event)}" if explained_event is not None else None
        ),
        honesty=honesty,
    )


def incident_record(item: IncidentFeedItem) -> IncidentRecord:
    """Encode the strict public item into the existing runtime incident row."""
    return IncidentRecord(
        incident_id=item.incident_id,
        state=item.state.value,
        created_at=item.opened_at,
        updated_at=item.updated_at,
        payload=item.model_dump(mode="json"),
    )


def incident_snapshot(
    records: tuple[IncidentRecord, ...],
    *,
    limit: int,
) -> IncidentFeedResponse:
    """Strictly revalidate durable rows before exposing any of them."""
    try:
        incidents = tuple(_item_from_record(record) for record in records)
    except (ValidationError, ValueError, TypeError) as exc:
        raise IncidentFeedDataError(f"stored incident snapshot is invalid: {exc}") from exc
    return IncidentFeedResponse(
        status="ready",
        incidents=incidents,
        count=len(incidents),
        limit=limit,
        detail=None,
    )


def unavailable_incident_snapshot(*, limit: int, detail: str) -> IncidentFeedResponse:
    """Return no partial rows when the live store cannot be trusted."""
    return IncidentFeedResponse(
        status="degraded",
        incidents=(),
        count=0,
        limit=limit,
        detail=_line(detail),
    )


def _item_from_record(record: IncidentRecord) -> IncidentFeedItem:
    item = IncidentFeedItem.model_validate_json(
        json.dumps(record.payload, allow_nan=False, separators=(",", ":"), sort_keys=True)
    )
    if item.incident_id != record.incident_id:
        raise ValueError("stored incident id disagrees with its payload")
    if item.state.value != record.state:
        raise ValueError("stored incident state disagrees with its payload")
    if item.opened_at != record.created_at or item.updated_at != record.updated_at:
        raise ValueError("stored incident timestamps disagree with its payload")
    return item


def _action_detail(action: DecisionAction) -> str:
    if action is DecisionAction.SUPPRESS:
        return "Decision suppressed; no production effect was requested."
    if action is DecisionAction.ALERT:
        return "Alert raised; no production effect was requested."
    if action is DecisionAction.ESCALATE_TO_HUMAN:
        return "Escalated to a person; no production effect was requested."
    if action is DecisionAction.AUTO_CONTAIN_THEN_ESCALATE:
        return "Containment authorized and escalation requested; no actuator outcome is attached."
    return "Action authorized by evidence; no actuator outcome is attached."


def _line(value: str) -> str:
    rendered = " ".join(value.split())
    if not rendered:
        raise ValueError("public incident text cannot be blank")
    return rendered
